[NativeAOT LLVM] Integrate dotnet.js into ILCompiler.LLVM

maraf opened the pull request:
- Include dotnet.js in ILCompiler framework folder
  - `dotnet.js` and `dotnet.native.js` built from typescript
  - `dotnet.es6.pre.js`, `dotnet.es6.lib.js` and `dotnet.es6.extpost.js` emscripten partials & library
- Remove emscripten build from `wasm.proj` and build it as part of naot-llvm build
- Produce dotnet.js compatible app when `DotNetJsApi=true` is set in users app
  - JS interop doesn't work yet
- Changes to typescript
  - Changes are driven by `NativeAOT` const for now. In follow ups we might come with a way to "override" functions different for NativeAOT
  - Use emscripten's `callMain` at the moment
  - Remove jiterpreter initialization
  - Environment variables not supported at the moment
  - Runtime options not supported
  - Remove managed exports initialization
- Add a smoke DotnetJs.csproj

maraf committed:
Fix typescript build

maraf committed:
Build wasmruntime and include dotnet.js in framework

maraf committed:
Include additional wasmruntime files in ILC framework

maraf committed:
Remove all cwraps. use dotnet.js when DotNetJsApi=true

maraf committed:
Make it work (TM)

maraf committed:
Extract DotNetJsApi link settings into a separate file

maraf committed:
Fix merge

maraf committed:
Remove emcc files from package

pavelsavara commented on src/mono/wasm/runtime/es6/dotnet.es6.pre.js:
```diff
@@ -2,5 +2,5 @@ if (_nativeModuleLoaded) throw new Error("Native module already loaded");
 // see https://github.com/emscripten-core/emscripten/issues/19832
 Module["getMemory"] = function () { return wasmMemory; }
-createDotnetRuntime = Module = createDotnetRuntime(Module);
+moduleArg = Module = moduleArg(Module);
```
why ?

maraf replied:
Emscripten 3.1.47 generates

```js
var createDotnetRuntime = (() => {
  var _scriptDir = import.meta.url;
  
  return (
async function(moduleArg = {}) {
```

I dunno yet if that can be configured somehow and I'm missing that flag or whether it's in empscripten (main has `3.1.34`)

pavelsavara replied:
<_EmccLinkFlags Include="-s EXPORT_NAME=&quot;'createDotnetRuntime'&quot;" />

pavelsavara replied:
maybe it's worth replacing the whole wasm.proj with latest from main ?

maraf replied:
> <_EmccLinkFlags Include="-s EXPORT_NAME="'createDotnetRuntime'"" />

I already have this one
https://github.com/dotnet/runtimelab/blob/1a55814195534969865a74693a9cf8cfb44ae776/src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets#L112
It produces the `createDotnetRuntime`, but the argument passed in is called `moduleArg`

maraf replied:
Here is a sample dotnet.native.js that is generated by emcripten 3.1.47 [dotnet.native.zip](https://github.com/dotnet/runtimelab/files/13603070/dotnet.native.zip)

pavelsavara replied:
hmmm, it seems we use `createDotnetRuntime` as default export in `dotnet.d.ts` but it seems it's a lie anyway.

pavelsavara commented on src/mono/wasm/runtime/cwraps.ts:
```diff
@@ -1,7 +1,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
-import MonoWasmThreads from "consts:monoWasmThreads";
+// import MonoWasmThreads from "consts:monoWasmThreads";
```
adding another "consts:NativeAOT" would solve a lot of this

pavelsavara commented on src/mono/wasm/runtime/invoke-cs.ts:
```diff
@@ -13,14 +13,14 @@ import {
 import { mono_wasm_new_external_root, mono_wasm_new_root } from "./roots";
 import { monoStringToString } from "./strings";
-import { MonoObjectRef, MonoStringRef, MonoString, MonoObject, MonoMethod, JSMarshalerArguments, JSFunctionSignature, BoundMarshalerToCs, BoundMarshalerToJs, VoidPtrNull, MonoObjectRefNull, MonoObjectNull, MarshalerType } from "./types/internal";
+import { MonoString, MonoMethod, JSMarshalerArguments, JSFunctionSignature, BoundMarshalerToCs, BoundMarshalerToJs, MarshalerType, MonoStringRef, MonoObject, MonoObjectRef, MonoObjectRefNull, VoidPtrNull, MonoObjectNull } from "./types/internal";
```
revert whole file ?

pavelsavara commented on src/mono/wasm/runtime/method-calls.ts:
```diff
@@ -1,652 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
this is legacy interop in an old directory. It seems this branch is really old ?

pavelsavara replied:
in the new layout you could completely disable legacy interop in rollup

maraf replied:
This is a previous merge damage. The file has moved to `src\mono\wasm\runtime\net6-legacy\method-calls.ts` and I'm just removing obsolete copy

maraf replied:
I have already disabled legacy interop

pavelsavara commented on src/mono/wasm/runtime/run.ts:
```diff
@@ -35,33 +32,35 @@ export async function mono_run_main_and_exit(main_assembly_name: string, args?:
 
-    return res;
+    // return res;
+    return (Module as any)["callMain"](args);
```
this will not work for async main

maraf replied:
I'm aware, temporal thing to be able to make sample that calls sync main. Proper solution is more complex

maraf committed:
Revert invoke-cs.ts

maraf committed:
Ensure WasmObjDir

maraf committed:
Build wasm runtime for debug config

maraf commented on eng/pipelines/runtimelab.yml:
```diff
@@ -123,9 +123,9 @@ extends:
                   librariesConfiguration: Release
             ${{ if eq(variables.isOfficialBuild, false) }}:
-              buildArgs: -s clr.aot+libs+nativeaot.packages -c $(_BuildConfig) /p:ArchiveTests=true
+              buildArgs: -s clr.aot+mono.wasmruntime+libs+nativeaot.packages -c $(_BuildConfig) /p:ArchiveTests=true
```
@yowl Do you know if `EMSDK` is set when targeting OSX or win-arm? https://dev.azure.com/dnceng-public/public/_build/results?buildId=492090&view=logs&j=752aa75d-2981-5712-2e83-4a413c44ec15&t=831180b0-4fa3-513d-132c-4972ab1af17e

SingleAccretion replied:
It's not. The CI is only set up to test/build WASM-related things for windows-[browser|wasi]. The other legs serve as sanity checks that we're not breaking upstream build/NAOT.

One way to fix this is to add `mono.wasmruntime` to `clr.aot`.

SingleAccretion commented:
A couple general questions, to start things off.

1\) What is the general code flow (dependency graph) of the JS code? For example, here's the detailed graph for the current runtime:
```
Emscripten-generated code (host)
  [Emscripten ABI]
    Native entry point from bootstrapper.cpp
      <application code>
      [Emscripten ABI]
        pal_random.js; --js-library
        emscripten_get_callstack; Emscripten-provided system function
```
This needn't be as detailed, I am just looking for what the major components are, what their responsibilities are, and where they live (/how the functions they consist of are called).

2\) What are the components in the build process of JS components? What is the build flow? How do you debug that build? Any common gotchas/known issues w.r.t. local development? We spend a very good chunk of our time fixing or altering the build in one way or another on this branch, so it's important to know how it works.

pavelsavara commented:
>         pal_random.js; --js-library

this doesn't exist anymore. It's moved into `src\mono\wasm\runtime\crypto.ts`

SingleAccretion commented:
> this doesn't exist anymore. It's moved into src\mono\wasm\runtime\crypto.ts

Yes; we restored it in this branch for the time being.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
This is clearly a copy from somewhere else. Is there a plan for un-copying it?

If there is not (which is reasonable, it's not easy to do), what is the plan to ensure it stays up-to-date with merges? How often do these compilation options change? How easy is to detect breaks from something missing?

If there is, it needs to be figured out how things that I would expect not to apply to NAOT can be segregated. There are a lot of them.

maraf replied:
I'm happy with the copy at the moment. I don't think these parameters change often. 

Original source of these values are `wasm.proj` and `WasmApp.Native.targets`. With mono we run emscripten with default configuration during runtime build and ship the outputs for usage without AOT (in this mode user doesn't need emscripten on his machine). We also ship the rsp files and json with configuration values. When "native build" (emscripten) is run during target app build, these default values are mixed with user defined parameters (applied in WasmApp.Native.targets).

I initially tried to mimic the same behavior (for parameter flow), but failed due to various reasons. I plan to come back to it and find a better way as we progress.

SingleAccretion replied:
Alright, if these don't change too often keeping a copy looks good to me. I will comment on the individual items then.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -96,13 +96,21 @@
+    <Copy SourceFiles="$(NativeOutputPath)$(TargetName).wasm" DestinationFolder="$(PublishDir)" Condition="'$(_targetOS)' == 'browser' and '$(DotNetJsApi)' != 'true'"/>
+    
+    <ItemGroup>
+      <_FilesToPublish Include="$(IlcFrameworkPath)\dotnet*.js" />
+      <_FilesToPublish Include="$(IlcFrameworkPath)\dotnet*.map" />
+      <_FilesToPublish Include="$(_WasmIntermediateOutputPath)dotnet.native.js" />
+      <_FilesToPublish Include="$(_WasmIntermediateOutputPath)dotnet.native.wasm" />
```
What does the output of a typical publish look like?

maraf replied:
Example output is here https://github.com/maraf/MinimalDotNetWasmNativeAOT/tree/dotnetjs/DotnetJsHack
- index.html & main.js is user code
- dotnet.js & dotnet.runtime.js are from runtime pack/ilc
  - .map are source maps pointing to typescript definitions on github
- dotnet.native.js & dotnet.native.wasm are emscripten outputs

So typically browser opens the index.html and that triggers main.js. User is then responsible for deciding if and how to start dotnet and what to invoke on it (it may or may not be main method)

SingleAccretion commented:
>  Add sample running main with dotnet.js

This should be added as a smoke test under `src/tests/nativeaot/SmokeTests/HelloWasm`.

maraf committed:
Add wasmruntime to clr.aot defaults

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+    <EmccExportedFunction Include="_fma" />
+    <EmccExportedFunction Include="_pow" />
+    <EmccExportedFunction Include="_fmodf" />
+    <EmccExportedFunction Include="_atan2f" />
+    <EmccExportedFunction Include="_fmaf" />
+    <EmccExportedFunction Include="_powf" />
+    <EmccExportedFunction Include="_asin" />
+    <EmccExportedFunction Include="_asinh" />
+    <EmccExportedFunction Include="_acos" />
+    <EmccExportedFunction Include="_acosh" />
+    <EmccExportedFunction Include="_atan" />
+    <EmccExportedFunction Include="_atanh" />
+    <EmccExportedFunction Include="_cbrt" />
+    <EmccExportedFunction Include="_cos" />
+    <EmccExportedFunction Include="_cosh" />
+    <EmccExportedFunction Include="_exp" />
+    <EmccExportedFunction Include="_log" />
+    <EmccExportedFunction Include="_log2" />
+    <EmccExportedFunction Include="_log10" />
+    <EmccExportedFunction Include="_sin" />
+    <EmccExportedFunction Include="_sinh" />
+    <EmccExportedFunction Include="_tan" />
+    <EmccExportedFunction Include="_tanh" />
+    <EmccExportedFunction Include="_asinf" />
+    <EmccExportedFunction Include="_asinhf" />
+    <EmccExportedFunction Include="_acosf" />
+    <EmccExportedFunction Include="_acoshf" />
+    <EmccExportedFunction Include="_atanf" />
+    <EmccExportedFunction Include="_atanhf" />
+    <EmccExportedFunction Include="_cbrtf" />
+    <EmccExportedFunction Include="_cosf" />
+    <EmccExportedFunction Include="_coshf" />
+    <EmccExportedFunction Include="_expf" />
+    <EmccExportedFunction Include="_logf" />
+    <EmccExportedFunction Include="_log2f" />
+    <EmccExportedFunction Include="_log10f" />
+    <EmccExportedFunction Include="_sinf" />
+    <EmccExportedFunction Include="_sinhf" />
+    <EmccExportedFunction Include="_tanf" />
+    <EmccExportedFunction Include="_tanhf" />
```
What are these math functions needed for? If they are not needed, they should be deleted.

maraf replied:
Removed, used by jiterpreter

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLinkRsp>$(_WasmIntermediateOutputPath)emcc-link.rsp</_EmccLinkRsp>
+    </PropertyGroup>
+    <PropertyGroup>
+      <_EmccExportedLibraryFunction>&quot;[@(EmccExportedLibraryFunction -> '%27%(Identity)%27', ',')]&quot;</_EmccExportedLibraryFunction>
+      <_EmccExportedRuntimeMethods>&quot;[@(EmccExportedRuntimeMethod -> '%27%(Identity)%27', ',')]&quot;</_EmccExportedRuntimeMethods>
+      <_EmccExportedFunctions>@(EmccExportedFunction -> '%(Identity)',',')</_EmccExportedFunctions>
```
Are these three user-facing item groups?

maraf replied:
Yes

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <EmccInitialHeapSize Condition="'$(EmccInitialHeapSize)' == ''">16777216</EmccInitialHeapSize>
+    </PropertyGroup>
+    <ItemGroup>
+      <EmscriptenEnvVars Include="DISABLE_LEGACY_JS_INTEROP=1" Condition="'$(WasmEnableLegacyJsInterop)' == 'false'" />
+      <EmscriptenEnvVars Include="DISABLE_LEGACY_JS_INTEROP=0" Condition="'$(WasmEnableLegacyJsInterop)' != 'false'" />
+      <EmscriptenEnvVars Include="WASM_ENABLE_SIMD=1" Condition="'$(WasmEnableSIMD)' != 'false'" />
+      <EmscriptenEnvVars Include="WASM_ENABLE_SIMD=0" Condition="'$(WasmEnableSIMD)' == 'false'" />
+      <EmscriptenEnvVars Include="WASM_ENABLE_EH=1" Condition="'$(WasmEnableExceptionHandling)' != 'false'" />
+      <EmscriptenEnvVars Include="WASM_ENABLE_EH=0" Condition="'$(WasmEnableExceptionHandling)' == 'false'" />
+      <EmscriptenEnvVars Include="ENABLE_AOT_PROFILER=$([System.Convert]::ToInt32($(WasmProfilers.Contains('aot'))))" />
+      <EmscriptenEnvVars Include="ENABLE_BROWSER_PROFILER=$([System.Convert]::ToInt32($(WasmProfilers.Contains('browser'))))" />
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=1" Condition="'$(RunAOTCompilation)' == 'true'" />
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=0" Condition="'$(RunAOTCompilation)' != 'true'" />
```
What do these environment variables affect?

maraf replied:
They are passed to dotnet.native.js and then to dotnet.runtime.js. They are probably not needed here, but it would add more changes to typescript

maraf replied:
We can reduce them to default values

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccExportedLibraryFunction>&quot;[@(EmccExportedLibraryFunction -> '%27%(Identity)%27', ',')]&quot;</_EmccExportedLibraryFunction>
+      <_EmccExportedRuntimeMethods>&quot;[@(EmccExportedRuntimeMethod -> '%27%(Identity)%27', ',')]&quot;</_EmccExportedRuntimeMethods>
+      <_EmccExportedFunctions>@(EmccExportedFunction -> '%(Identity)',',')</_EmccExportedFunctions>
+      <_EmccDefaultFlags Condition="'$(WasmEnableSIMD)' == 'true'">-msimd128</_EmccDefaultFlags>
+      <EmccStackSize Condition="'$(EmccStackSize)' == ''">5MB</EmccStackSize>
+      <EmccInitialHeapSize Condition="'$(EmccInitialHeapSize)' == '' and '$(_WasmCalculatedInitialHeapSize)' != '' and $(_WasmCalculatedInitialHeapSize) > 16777216">$(_WasmCalculatedInitialHeapSize)</EmccInitialHeapSize>
+      <EmccInitialHeapSize Condition="'$(EmccInitialHeapSize)' == ''">16777216</EmccInitialHeapSize>
```
There are a few things that already exist in our targets (e. g. we set the default stack size). Let's rationalize things as follows: there are "JS API" properties and "runtime" properties.

"JS API" properties apply to Browser + `'$(DotNetJsApi)'=='true'` only. They affect JS-specific things and/or expose `emcc` knobs. They should be used/defined here.

"Runtime" properties apply to both Browser and WASI. They will often affect core runtime semantics. They will be used/defined here: https://github.com/dotnet/runtimelab/blob/6a2d69f4ac6bce0ff441806172c840f08ed2e952/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L504-L563 (for now, we're planning to move them to a different file).

In this framework:
1) `EmccStackSize` is a "runtime" property. It should affect `IlcWasmStackSize`.
2) `EmccInitialHeapSize` - likewise.
3) `WasmEnableSIMD` is a no-op - we don't support SIMD yet - and so should be deleted.

I am also interested which of these MSBuild properties are documented. There should a clear separation between what is public (and cannot be changed), like `WasmEnableSIMD`, and what is not.

I will make a number of comments below about "runtime" properties. The way to fix them is to either: move the linker flag setting "down", or add an NYI comment instead, or just delete it because it's not a public property and we don't need to support all `emcc` flags by hand.

maraf replied:
I'll move runtime properties as suggested
Removed `WasmEnableSIMD`

maraf replied:
All public props are documented here https://github.com/dotnet/runtime/blob/e9ddf98c8502cfa00823dc94d3fcbfe5aa03ece7/src/mono/wasm/build/WasmApp.Common.targets#L6

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <EmscriptenEnvVars Include="ENABLE_BROWSER_PROFILER=$([System.Convert]::ToInt32($(WasmProfilers.Contains('browser'))))" />
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=1" Condition="'$(RunAOTCompilation)' == 'true'" />
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=0" Condition="'$(RunAOTCompilation)' != 'true'" />
+      <_EmccCommonFlags Include="$(_DefaultEmccFlags)" />
```
What sets this item group?

maraf replied:
The `_DefaultEmccFlags` is empty here, removed

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=1" Condition="'$(RunAOTCompilation)' == 'true'" />
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=0" Condition="'$(RunAOTCompilation)' != 'true'" />
+      <_EmccCommonFlags Include="$(_DefaultEmccFlags)" />
+      <_EmccCommonFlags Include="$(EmccFlags)" />
```
What/who sets this property?

maraf replied:
This is a public property. Users can set it

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <EmscriptenEnvVars Include="RUN_AOT_COMPILATION=0" Condition="'$(RunAOTCompilation)' != 'true'" />
+      <_EmccCommonFlags Include="$(_DefaultEmccFlags)" />
+      <_EmccCommonFlags Include="$(EmccFlags)" />
+      <_EmccCommonFlags Include="-s MAXIMUM_MEMORY=$(EmccMaximumHeapSize)" Condition="'$(EmccMaximumHeapSize)' != ''" />
```
This is a runtime property.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccCommonFlags Include="$(_DefaultEmccFlags)" />
+      <_EmccCommonFlags Include="$(EmccFlags)" />
+      <_EmccCommonFlags Include="-s MAXIMUM_MEMORY=$(EmccMaximumHeapSize)" Condition="'$(EmccMaximumHeapSize)' != ''" />
+      <_EmccLDFlags Include="-s EXPORT_ES6=1" />
+      <_EmccLDFlags Include="-s INVOKE_RUN=0" />
+      <_EmccLDFlags Include="-s ALLOW_TABLE_GROWTH=1" />
+      <_EmccLDFlags Include="-s FORCE_FILESYSTEM=1" />
+      <_EmccLDFlags Include="-s EXPORT_NAME=&quot;'createDotnetRuntime'&quot;" />
```
These are required for the core functionality, correct? If so, there should be a comment to that effect.

maraf replied:
Most of the settings here are required to make it functional. What is the goal you would like to achieve?

SingleAccretion replied:
Understand (for each property), why it is there.

For some it is obvious (some functions need to be exported because JS/TS uses them), for others - no (`FORCE_FILESYSTEM=1`, `ALLOW_TABLE_GROWTH=1`, for example).

maraf replied:
I see. I'll add comments to various settings

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLDFlags Include="-s ENVIRONMENT=&quot;'web,webview,worker,node,shell'&quot;" />
+      <_EmccLDFlags Include="-s MODULARIZE=1" />
+      <_EmccLinkStepArgs Include="@(_EmccLDFlags)" />
+      <_EmccLDFlags Include="$(EmccLinkOptimizationFlag)" />
```
Runtime property.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLDFlags Include="$(EmccLinkOptimizationFlag)" />
+      <_EmccLDFlags Include="@(_EmccCommonFlags)" />
+      <_EmccLDSFlags Include="-s INITIAL_MEMORY=$(EmccInitialHeapSize)" />
+      <_EmccLDSFlags Include="-s STACK_SIZE=$(EmccStackSize)" />
```
Runtime properties.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLDFlags Include="-s ALLOW_TABLE_GROWTH=1" />
+      <_EmccLDFlags Include="-s FORCE_FILESYSTEM=1" />
+      <_EmccLDFlags Include="-s EXPORT_NAME=&quot;'createDotnetRuntime'&quot;" />
+      <_EmccLDFlags Include="-s ENVIRONMENT=&quot;'web,webview,worker,node,shell'&quot;" />
```
What does this do?

maraf replied:
IIRC these should eliminate emscripten code for other environments? cc @pavelsavara

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLDSFlags Include="-s INITIAL_MEMORY=$(EmccInitialHeapSize)" />
+      <_EmccLDSFlags Include="-s STACK_SIZE=$(EmccStackSize)" />
+      <_EmccLDSFlags Condition="'$(EmccEnvironment)' != ''" Include="-s ENVIRONMENT=&quot;$(EmccEnvironment)&quot;" />
+      <_EmccLDSFlags Condition="'$(EmccEnableAssertions)' == 'true'" Include="-s ASSERTIONS=1" />
```
Is this a public property?

maraf replied:
Yes

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_WasmExtraJSFile Include="$(IlcFrameworkNativePath)src\*.%(JSFileType.Identity)"     Kind="%(JSFileType.Kind)" />
+      <_WasmExtraJSFile Include="$(IlcFrameworkNativePath)src\es6\*.%(JSFileType.Identity)" Kind="%(JSFileType.Kind)" />
+      <_EmccLinkStepArgs Include="--%(_WasmExtraJSFile.Kind) &quot;%(_WasmExtraJSFile.Identity)&quot;" Condition="'%(_WasmExtraJSFile.Kind)' != ''" />
+      <_EmccLinkStepArgs Include="-o &quot;$(_WasmIntermediateOutputPath)dotnet.native.js&quot;" />
```
This should set `$(NativeBinary)` instead of overriding things on the linker command line.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLinkStepArgs Include="-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=$(_EmccExportedLibraryFunction)" Condition="'$(_EmccExportedLibraryFunction)' != ''" />
+      <_EmccLinkStepArgs Include="-s EXPORTED_RUNTIME_METHODS=$(_EmccExportedRuntimeMethods)" />
+      <_EmccLinkStepArgs Include="-s EXPORTED_FUNCTIONS=$(_EmccExportedFunctions)" />
+      <_EmccLinkStepArgs Include="--source-map-base http://example.com" />
```
> `http://example.com`

Ehhm, how does this work?

maraf replied:
Removed

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+      <_EmccLinkStepArgs Include="-s EXPORTED_RUNTIME_METHODS=$(_EmccExportedRuntimeMethods)" />
+      <_EmccLinkStepArgs Include="-s EXPORTED_FUNCTIONS=$(_EmccExportedFunctions)" />
+      <_EmccLinkStepArgs Include="--source-map-base http://example.com" />
+      <_EmccLinkStepArgs Include="-Wno-limited-postlink-optimizations" />
```
```suggestion
```
We don't run `wasm-opt` by default so this is not useful (see https://github.com/dotnet/runtimelab/issues/2357).

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,146 @@
+    </ItemGroup>
+    <WriteLinesToFile Lines="@(_EmccLinkStepArgs)" File="$(_EmccLinkRsp)" Overwrite="true" WriteOnlyWhenDifferent="true" />
+    <PropertyGroup>
+      <EmccExtraArgs>$(EmccExtraArgs) $(_EmccDefaultFlags) "@$(_EmccLinkRsp)"</EmccExtraArgs>
```
`EmccExtraArgs` is a legacy flag that should not be used. `LinkerArg` item group should be used instead. We want all of the linker setting to be contained in one response file in the end.

maraf replied:
IIRC I have some troubles when everything was in single file, At least with `EXPORTED_FUNCTIONS`. I'll look at it to be sure

maraf replied:
The problem was only with ordering. I have succeeded after declaring a new `LinkNativeLlvmDependsOn` and appending DotNetJsApi settings after applying "default" settings

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -583,4 +585,14 @@ The .NET Foundation licenses this file to you under the MIT license.
   <Import Project="$(MSBuildThisFileDirectory)Microsoft.NETCore.Native.Publish.targets" Condition="'$(NativeCompilationDuringPublish)' != 'false'" />
 
+
+  <UsingTask TaskName="ReadEmccProps" AssemblyFile="$(IlcBuildTasksPath)"
```
> `ReadEmccProps`

Where is this task used (I can't find the reference)?

maraf replied:
My mistake

maraf committed:
Reduce EmscriptenEnvVars to default values

maraf committed:
Remove _DefaultEmccFlags, no-limited-postlink-optimizations

maraf committed:
Remove ReadEmccProps

pavelsavara commented:
FYI https://github.com/dotnet/runtime/pull/95775 we are refactoring those proj quite often. 
I'm not sure that optimizing it here makes ton of sense from merging perspective.

SingleAccretion commented:
>  I'm not sure that optimizing it here makes ton of sense from merging perspective.

Since we decided to keep a copy (of the compiler options-related MSBuild things), there is no conflict risk. There is still the problem of keeping the copy up-to-date, but it's the same whether we simplify it or not. And a number of these options have to be rewritten in any case.

maraf committed:
Fix reading dotnet.js from RuntimeConfiguration outputs instead of LIbraryConfiguration outputs

maraf committed:
Fix for the previous "Fix reading dotnet.js from RuntimeConfiguration outputs instead of LIbraryConfiguration outputs"

maraf committed:
Use constant to disable mono cwraps

maraf committed:
Include js files only from es6 folder

maraf committed:
Integrate copy files into build

maraf committed:
Copy WasmExtraFilesToDeploy to native and publish dirs

maraf committed:
Fix using args from withApplicationArguments

maraf committed:
Clean up linkers args (1)

maraf committed:
Some test with dotnet.js API

maraf committed:
Clean up linkers args (2)

maraf committed:
Remove math functions used by jiterpreter

maraf committed:
Distribute "JS API" and "Runtime" props

maraf committed:
Use single link rsp file

Declare a new LinkNativeLlvmDependsOn and append DotNetJsApi settings after applying "default" settings

maraf committed:
Default config for native aot

maraf commented on src/mono/wasm/runtime/run.ts:
```diff
@@ -48,20 +45,22 @@ export async function mono_run_main(main_assembly_name: string, args?: string[])
         }
     }
+    
+    // TODO MF: Fix mono_run_main
```
@pavelsavara  Do you know of a way to "conditionaly import from different file in typescript"? Something like what we can accomplish in C# with partial classes + conditional file include in csproj?

One way to solve this in JS/TS world I can think of is by going OOP. Declare interfaces/helper objects for things we might need to swap in NAOT, but that is slightly out of the scope for this PR.

pavelsavara replied:
you are talking about 2 different concepts, I think.
1) generating different code
2) taking another code path while both are available at the runtime. 

Re: 1)
I strongly prefer this option, because of the download size.
We have ways how to achieve that with RollupJS, I think.

```ts
import IsNativeAOT from "consts:NativeAOT";

export const myFunction = IsNativeAOT
    ? function myNaotFunction() { return null }
    : function myMonoFunction() { return null }
```

Would this work ?

Ideally that should be in some top level file, but our TS is bit of spaghetti WRT file dependencies.
At the moment those are
- loader/index.ts
- exports.ts
- exports-linker.ts
- types/export-types.ts - you can ignore types I guess
- net6-legacy/export-types.ts - you can ignore types I guess

Re 2)
- let's not do classes in low level runtime JS/TS. Perhaps interfaces, but do we really need that instead of simple `if` condition in few places calling the other implementation ? How many methods you need to differentiate ?

pavelsavara replied:
Obviously `myNaotFunction` could be in different file than `myMonoFunction`. The fact you include both files doesn't matter because tree shaking will remove it.

maraf replied:
I don't want to have them both on runtime. I was referring to a similar pattern you mentioned with the if, but in a way that only runtimelab branch can define the overrides. Something like that upstream have `api.myFunction = monoFunction` and here we can override it `api.myFunction = naotFunction`.

maraf replied:
I probably can already override some of the methods with loaderHelpers object, like invoking main. I will need to override JS marshalling functions in follow up

maraf committed:
Replace comments with NativeAOT constant

maraf committed:
Add main.js as another expected test entrypoint?

maraf committed:
Fix typescript imports for NativeAOT

maraf committed:
Fix mono_exit

maraf committed:
Add package.json to test to support modules on node. Exit from node.

SingleAccretion commented on src/tests/nativeaot/SmokeTests/DotnetJs/Program.cs:
```diff
@@ -0,0 +1,11 @@
+    static int Main(string[] args)
+    {
+        Console.WriteLine("Hello, DotnetJs!");
+        Console.WriteLine($"Args {String.Join(", ", args)}");
```
Validate that the passed-in arguments are as expected?

SingleAccretion commented on src/tests/nativeaot/SmokeTests/DotnetJs/wwwroot/package.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "type": "module"
+}
```
Is this needed because of the "modern JS" use? I remember Node accepts `.mjs` files for it by default.

maraf replied:
Yes, that's because of use of ES6 modules. `.mjs` is alternative solution. I'm not if that requires to have all ES6 modules as `.mjs` (including dotnet.js)

pavelsavara replied:
It doesn't requite that from other than top level module. But older versions of nodejs don't respect `.mjs` well.

SingleAccretion commented on src/tests/nativeaot/SmokeTests/DotnetJs/wwwroot/main.js:
```diff
@@ -0,0 +1,12 @@
+
+import { dotnet, exit } from './dotnet.js'
+
+const { runMain } = await dotnet
+    .withApplicationArguments("A", "B", "C")
+    .create();
```
Generated Emscripten code will handle passing arguments passed to Node automatically. To do the same here would require something like:
```
if (Node) .withApplicationArguments(getCmdArgs()) else { ... }
```
right?

pavelsavara replied:
nope, we do it for you too.
https://github.com/dotnet/runtime/blob/1d4897d29655f93677240fc93a1626a6e334451d/src/mono/wasm/runtime/run.ts#L41-L50

SingleAccretion commented on src/tests/nativeaot/SmokeTests/DotnetJs/DotnetJs.csproj:
```diff
@@ -0,0 +1,17 @@
+    <!-- IL3049 warns about mismatched signatures; we are intentionally testing they are handled correctly
+         when the definition is statically linked -->
+    <NoWarn>$(NoWarn);IL3049</NoWarn>
+
```
```suggestion
```

SingleAccretion commented on src/tests/Common/scripts/nativeaottest.cmd:
```diff
@@ -21,6 +21,10 @@ if not exist %__JsFilePath% (
   set __JsFilePath=%__JsFilePath:~0,-3%.mjs
 )
 
+if not exist %__JsFilePath% (
+  set __JsFilePath=%1\native\main.js
+)
```
I am a bit curious on the thinking about allowing project-specific output file names. It is non-standard for .NET projects for the output binary to not be named after the project itself.

maraf replied:
It is non standard for .NET projects. On web it's up to the user how he names his entrypoint. Nothing settled IMO. Our templates has main.js as entrypoint.

I can rename main.js in the test to DotnetJs.js (as the project itself) if you prefer that

pavelsavara replied:
`bike_shed.js` 😉

SingleAccretion replied:
No, it's fine to leave this as it is for now. I suppose it is not really the `main` file I am worried about, but the all of `dotnet.*` files. Ideally, the target there would be to have the same output as what Emscripten generates: one `Project.js`, one `Project.wasm`. It helps optics/perception. Nothing to do about that at this point, though.

pavelsavara replied:
We have multiple files for performance and modularity reasons. We are not going back to single JS file. It's more trouble than worth to live inside emscripten's JS postprocessing tooling. Also you need extra JS file for threads.

pavelsavara replied:
Actually I think we will have more modules, so that we could load some of the larger JS components only when necessary. Download size and startup speed are tier 1) goal for us.

pavelsavara replied:
We could discuss more here https://github.com/dotnet/runtime/issues/86162

SingleAccretion replied:
I would hope one of the goals of future designs is being minimal in the JS surface area (i. e. expose only the absolutely necessary platform APIs and do actual work in managed code). This works better with trimming. I understand with the current state of things it is a challenging goal due to high interop costs.

I think we both agree on the fact that adding one more `.js` file to the output has a certain (not insignificant) cost, it's not like adding one more `.cs` file to the source tree, which is essentially free.

pavelsavara replied:
No, I don't think doing most it in C# is possible. But I'm open to see specific improvements.

pavelsavara replied:
Ideally on main branch.

pavelsavara replied:
> No, I don't think doing most it in C# is possible. But I'm open to see specific improvements.

I think I was bit blunt here, let me explain.
- big part of the JS/TS we have deals with download of assets and startup. That can't be C#, chicken/egg. It's heavily async, using network badwidth and CPU at the same time.
- the implementation of the JS interop. chicken/egg.
- jiterpreter, also kind of chicken/egg. Maybe this could become C. But TS is easier to maintain.
- whatever emscripten/POSIX is not good enough for us. Hacks into emscripten JS internals, small code size
- timers, single-thread "thread pool", `Task` asynchrony all of which has to play well with browser event loop.
- fancy trade-offs, like hybrid globalization, using browser's APIs instead of ICU. There is very brittle balance of perf vs size. How you pass buffers makes huge difference. Maybe this could be a candidate for moving it to low level C# with raw pointers etc. [Example ](https://github.com/dotnet/runtime/pull/95583)of how we are working on good tradeoff right now.
- http and WS clients. There is thin TS code that wraps the browser API in a way that makes it possible to consume as JS interop. Dealing with buffers and lack cancellation tokens in JS.

Well, but perhaps you could bring fresh perspective @SingleAccretion :)

pavelsavara replied:
jiterpreter is something that could become separate ES6 module, or mono C code or RollupJS conditional compilation. Because it only takes bytes for NativeAOT right now.

SingleAccretion replied:
I agree it is a hard problem with many angles. My perspective is "pay for play" design. One of the things I like about Emscripten is that it tries hard to square JS with the circle of C-style static linking - it means that if you don't use "feature X", the code for "feature X" gets dropped. E. g. if jiterpreter was written in C, it would automatically get linked out because it's unused (I am not saying that it should have been written in C, just an example). This sort of trimming is the core reason NAOT apps can be small (recall, after all, that IL is more compact than native code, and more compact than WASM), and become huge if, say, [something brings along half of WPF in a WinForms app](https://github.com/dotnet/runtime/pull/91988)...

I am not sure how trimmable TS is (yet). You can do better than "just include everything" by separating things into modules, but for best results (e. g.: only link in JS `string` marshalling if something has a JSImport/JSExport with a `string` in it), you need function-level granularity.

pavelsavara replied:
JS is well trimmable, but we don't have things like rollupjs in the workload/emscripten bundle. So we don't have the tools on customer machine, where he sets the feature flags. Maybe one day we will add rollupjs to the workload and shake the tree some more. 

emscripten makes JS trimmable even together with C code, but it only works if you are willing to write your JS in C style. (no closures). It also damages your code if you use modern ES syntax. We are not willing to try again. If you want to have look, see Net6 codebase we had.

pavelsavara replied:
> My perspective is "pay for play" design. 

We also prefer that.

> (e. g.: only link in JS `string` marshalling if something has a JSImport/JSExport with a `string` in it)

linker across C#/C/JS is though. Especially with JS being very dynamic. 

In this specific example, we interpret the JS side of the "marshal single argument" from metadata about the method signature. And so we have to have all marshallers available on JS side. 

Alternative is to ship generated JS code for each JSImport/JSExport. It has tooling and deployment troubles. Roslyn doesn't support that either.

SingleAccretion replied:
> We are not willing to try again. If you want to have look, see Net6 codebase we had.

Just curious - was it 'simply' that the code was ugly, or there was some more fundamental problem (too many Emscripten bugs, hard to track down failure modes, ...)?

> In this specific example, we interpret the JS side of the "marshal single argument" from metadata about the method signature. And so we have to have all marshallers available on JS side.

Its possible to write it in such a way so as to make the dependencies explicit. For example, you can imagine that for:
```
[JSImport] void func(string);
```
The generator would generate:
```
[DllImport("*")] void js_marshal_string();
[JSImport, DependsOn("js_marshal_string")] void func(string); // Imaginary attribute; just for the sake of argument
```
That imaginary attribute would lead to passing `--undefined:js_marshal_string` to the linker, which would preserve that marshalling method in a C-like static linking scenario. Admittedly, this is questionable design, but good enough for illustrative purposes.

It is fine to say that "well, this is a rather complicated thing to do, so we will settle for some middle-of-the-road design, like making trimming based in feature switches and JS modules (one module per switch)", but then you really have to be careful with the amount of JS (hence my earlier point about avoiding it).

maraf replied:
> Alternative is to ship generated JS code for each JSImport/JSExport. It has tooling and deployment troubles. Roslyn doesn't support that either.

I'm dreaming about this one ❤️

pavelsavara replied:
> Just curious - was it 'simply' that the code was ugly, or there was some more fundamental problem (too many Emscripten bugs, hard to track down failure modes, ...)?

Just go and have look :)

> [DllImport("*")] void js_marshal_string();
> [JSImport, DependsOn("js_marshal_string")] void func(string); // Imaginary attribute; just for the sake of argument

And to use that to fabricate roots for JS trimmer, right ?
`JSExport` could be more difficult.
Would that be multi-pass ?

> It is fine to say that "well, this is a rather complicated thing to do, so we will settle for some middle-of-the-road design, like making trimming based in feature switches and JS modules (one module per switch)", but then you really have to be careful with the amount of JS 

That's exactly where we are.

> (hence my earlier point about avoiding it).

In that sense I agree, but not as top priority or goal on it's own.

SingleAccretion replied:
> And to use that to fabricate roots for JS trimmer, right ? JSExport could be more difficult.

Right. I haven't thought about this too deeply. On the surface, both imports and exports have similar requirements, "just" the direction of marshaling is reversed.

>  Would that be multi-pass ?

I don't immediately see why it needs to be.

SingleAccretion commented on src/tests/Common/dirs.proj:
```diff
@@ -15,6 +15,7 @@
       <DisabledProjects Include="$(TestRoot)Performance\Scenario\JitBench\unofficial_dotnet\JitBench.csproj" /> <!-- no official build support for SDK-style netcoreapp2.0 projects -->
       <DisabledProjects Include="$(TestRoot)TestWrappers*\**\*.csproj" />
       <DisabledProjects Include="$(TestRoot)nativeaot\SmokeTests\HelloWasm\*.csproj" Condition="!('$(TestBuildMode)' == 'nativeaot' and '$(TargetArchitecture)' == 'wasm')"/>
+      <DisabledProjects Include="$(TestRoot)nativeaot\SmokeTests\DotNetJs\*.csproj" Condition="!('$(TestBuildMode)' == 'nativeaot' and '$(TargetArchitecture)' == 'wasm')"/>
```
Instead of this `DisabledProjects` mechanism, the preferred way to disabled individual projects is in their `.csproj`:
```xml
<CLRTestTargetUnsupported Condition="'$(TargetsBrowser)' != 'true'">true</CLRTestTargetUnsupported>
```
(`TargetBrowser` since this won't run on WASI)

SingleAccretion commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -29,14 +29,15 @@ import { mono_log_debug, mono_log_error, mono_log_warn, mono_set_thread_id } fro
 import { localHeapViewU8 } from "./memory";
 import { assertNoProxies } from "./gc-handles";
+import { jiterpreter_allocate_tables } from "./jiterpreter-support";
+import { init_managed_exports } from "./managed-exports";
```
Import order swap intentional?

maraf replied:
No, thanks!

SingleAccretion commented on src/mono/wasm/wasm.proj:
```diff
@@ -3,6 +3,8 @@
   <UsingTask TaskName="Microsoft.WebAssembly.Build.Tasks.RunWithEmSdkEnv" AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
 
   <PropertyGroup>
+    <EMSDK_PATH Condition="'$(EMSDK_PATH)' == ''">$(EMSDK)</EMSDK_PATH>
```
These `wasm.proj` changes need to be rationalized differently, essentially such that you can still build its upstream version as-is.

I would expect it to be along the lines of: add a wrapper project to some `/nativeaot` subdirectory, which would configure it as "build for NAOT", and either invoke `<MSBuild />` or just `<Import Project="wasm.proj" />` (I don't recall if you can do the latter with such projects). The end result should be that the output ends up under `artifacts/bin/coreclr/wasm.Browser.Debug/aotsdk`.

Then add that MSBuild project as a separate subset, let's say `clr.aotwasmruntime`, and include it in `clr.aot` for WASM.

maraf replied:
What are the benefits at this stage? Are you using the wasmruntime packages produced from this branch? I would rather put my effort into making more things working and this seems to be more important in longer term

SingleAccretion replied:
> What are the benefits at this stage?

It's the same as with typescript code - less diff from upstream in upstream files. Each diff is a maintainace overhead.

Also, it has been our policy to not break "upstream build". There are two reasons for this:
a) Debugging broken build is much easier when you have a working baseline, by following the "what is it that we do different" lines of logic.
b) It protects against designs which would not work as an upstreamable change.

maraf replied:
I see. I don't think I can make it without at least adding some conditions in wasm.proj. Do you prefer to have such changes or should I just duplicate the wasm.proj?

SingleAccretion replied:
> don't think I can make it without at least adding some conditions in wasm.proj. Do you prefer to have such changes?

Yes; it looks like there shouldn't be that many.

maraf replied:
Would be enough to add conditions in wasm.proj based on `PrimaryRuntimeFlavor == CoreCLR`?

SingleAccretion replied:
I don't think it should be `PrimaryRuntimeFlavor` (this really should be `Mono` for WASM, but isn't for historical reasons), but just `RuntimeFlavor` should/could work. The separate project is not strictly necessary, I was just thinking it could help to reduce the number of conditions required. If it's not necessary that's all the better.

maraf replied:
In https://github.com/dotnet/runtimelab/pull/2453/commits/993551519a976c59cec5d89d758527cc4ef5be40 I have tried to make the changes to wasm.proj smallest possible. I don't see how new project would minify them, but I'm open to suggestions

SingleAccretion replied:
`PrimaryRuntimeFlavor` -> `RuntimeFlavor`.

<details>
<summary>Why do we have these two properties?</summary>

The "primary" runtime flavor was added some time ago to support builds of CoreCLR components like `ilasm` on platforms that otherwise use Mono as the default runtime. It is supposed to be a pure function of the target platform, i. e. `./build clr.blabla -a UsuallyAMonoTarget` will have `PrimaryRuntimeFlavor` set to `Mono`. `RuntimeFlavor` is based on the component you're building and should be `CoreCLR` (one may ask: what happens if you build both? Well, _something_ does...).
</details>

maraf committed:
Fix result code

SingleAccretion commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -526,23 +529,27 @@ async function mono_wasm_before_memory_snapshot() {
-            mono_wasm_setenv(k, v);
-        else
-            throw new Error(`Expected environment variable '${k}' to be a string but it was ${typeof v}: '${v}'`);
+    if (!NativeAOT) {
+        for (const k in runtimeHelpers.config.environmentVariables) {
+            const v = runtimeHelpers.config.environmentVariables![k];
+            if (typeof (v) === "string")
+                mono_wasm_setenv(k, v);
+            else
+                throw new Error(`Expected environment variable '${k}' to be a string but it was ${typeof v}: '${v}'`);
+        }
+        if (runtimeHelpers.config.runtimeOptions)
+            mono_wasm_set_runtime_options(runtimeHelpers.config.runtimeOptions);
```
Unfortunately, (default) git merge is not the greatest at dealing with conflicts that involve indentation changes. Can we do a little trick like:
```
if (NativeAOT) runtimeHelpers.config.environmentVariables = null;
```
To reduce the diff?

SingleAccretion replied:
Separate request: could you list things which "need to be fixed for NAOT" (such as this environment variable support) in #2434?

maraf replied:
> Unfortunately, (default) git merge is not the greatest at dealing with conflicts that involve indentation changes. Can we do a little trick like:
> 
> ```
> if (NativeAOT) runtimeHelpers.config.environmentVariables = null;
> ```
> 
> To reduce the diff?

I see. I'll try to do as many as possible edition as append only

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -582,5 +600,4 @@ The .NET Foundation licenses this file to you under the MIT license.
   </Target>
 
   <Import Project="$(MSBuildThisFileDirectory)Microsoft.NETCore.Native.Publish.targets" Condition="'$(NativeCompilationDuringPublish)' != 'false'" />
-
```
Unintentional whitespace deletion.

SingleAccretion commented on src/mono/wasm/runtime/loader/run.ts:
```diff
@@ -477,9 +478,13 @@ async function initializeModules(es6Modules: [RuntimeModuleExportsInternal, Nati
+    } else {
+        if (!emscriptenModule.configSrc && (!loaderHelpers.config || Object.keys(loaderHelpers.config).length === 0 || (!loaderHelpers.config.assets && !loaderHelpers.config.resources))) {
+            // if config file location nor assets are provided
+            emscriptenModule.configSrc = "./blazor.boot.json";
```
Same question about avoiding indentation changes.

SingleAccretion commented on src/tests/nativeaot/SmokeTests/DotnetJs/wwwroot/main.js:
```diff
@@ -0,0 +1,12 @@
+
+var result = await runMain();
+console.log(`Exit code ${result}`);
+exit(result == 42 ? 100 : 1);
```
It would be more direct to return the exist from from the test (with appropriate changes to the test).
```suggestion
var result = await runMain();
console.log(`Exit code ${result}`);
exit(result);
```

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,94 @@
+  </Target>
+  <Target Name="CopyDotnetJsAfterLinking" AfterTargets="LinkNativeLlvm">
+    <ItemGroup>
+      <_FilesToCopyToNative Include="$(IlcFrameworkNativePath)\dotnet*.js" />
+      <_FilesToCopyToNative Include="$(IlcFrameworkNativePath)\dotnet*.map" Condition="'$(WasmEmitSourceMap)' == 'true'" />
+      <_FilesToCopyToNative Include="@(WasmExtraFilesToDeploy)" /><!-- Use defined extra files to be included in the build output -->
```
Can the same effect be achieved by using standard items:
```
<None Include="$(IlcFrameworkNativePath)\dotnet*.js" CopyToOutputDirectory="true" CopyToPublishDirectory="true"  />
```
?

It would remove the duplication between this and publishing targets below.

maraf replied:
That doesn't work

SingleAccretion replied:
Do you know why (perhaps a binlog I could take a look at)?

maraf replied:
I can reuse the collection of `_FilesToCopyToNative` in publish. Do you have a better suggestion?

SingleAccretion replied:
Works for me.

SingleAccretion reviewed (commented):
A bit more feedback. Major point is the `wasm.proj` build. The [earlier](https://github.com/dotnet/runtimelab/pull/2453#issuecomment-1847082070) two questions still stand as well.

Thank you for adding the comments!

maraf committed:
Update src/tests/nativeaot/SmokeTests/DotnetJs/DotnetJs.csproj

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

maraf committed:
Check args. More straightforward exit code

maraf committed:
Revert import swap

maraf committed:
Revert removed empty line

maraf commented:
> 1) What is the general code flow (dependency graph) of the JS code? 

In browser case and the new smoke test concretely, it is
- Browse opens index.html
- User `main.js` script executes
  - Imports `dotnet.js` which is public API for .NET
  - Calls `dotnet.create`
    - Loads JS modules (`dotnet.native.js` and `dotnet.runtime.js`)
    - Loads WASM (`dotnet.native.wasm`)
    - Emscripten module starts
  - Calls `runMain`
    - Emscripten's `callMain` executes
 
> 2) What are the components in the build process of JS components? What is the build flow? How do you debug that build? Any common gotchas/known issues w.r.t. local development? We spend a very good chunk of our time fixing or altering the build in one way or another on this branch, so it's important to know how it works.

Typescript is built using rollup. In debug build we get unminified version. In release they are minified. In both cases there are source maps pointing to original typescript definitions.

Does it answer the questions?

SingleAccretion commented:
> Does it answer the questions?

Yes, thank you.

maraf committed:
Rewrite typescript changes as append only

maraf committed:
Rewrite typescript changes as append only (2)

maraf committed:
Use CLRTestTargetUnsupported

maraf committed:
Reuse _FilesToCopyToNative in publish

maraf committed:
Minimize changes to wasm.proj

SingleAccretion commented on src/mono/wasm/runtime/invoke-cs.ts:
```diff
@@ -13,14 +13,14 @@ import {
 import { mono_wasm_new_external_root, mono_wasm_new_root } from "./roots";
 import { monoStringToString } from "./strings";
-import { MonoObjectRef, MonoStringRef, MonoString, MonoObject, MonoMethod, JSMarshalerArguments, JSFunctionSignature, BoundMarshalerToCs, BoundMarshalerToJs, VoidPtrNull, MonoObjectRefNull, MonoObjectNull, MarshalerType } from "./types/internal";
+import { MonoString, MonoMethod, JSMarshalerArguments, JSFunctionSignature, BoundMarshalerToCs, BoundMarshalerToJs, MarshalerType, MonoStringRef, MonoObject, MonoObjectRef, MonoObjectRefNull, VoidPtrNull, MonoObjectNull } from "./types/internal";
```
Looks like this file wasn't reverted?

maraf replied:
My bad, thanks

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/DotNetJsApi.targets:
```diff
@@ -0,0 +1,94 @@
+  <Target Name="PrepareDotNetJsApiForLinking">
+    <PropertyGroup>
+      <_WasmIntermediateOutputPath Condition="'$(WasmBuildingForNestedPublish)' == ''">$([MSBuild]::NormalizeDirectory($(IntermediateOutputPath), 'wasm'))</_WasmIntermediateOutputPath>
+      <_EmccLinkRsp>$(_WasmIntermediateOutputPath)emcc-link.rsp</_EmccLinkRsp>
```
Both look unused.

maraf committed:
Guard mono_wasm_set_runtime_options for naot

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -491,14 +492,17 @@ The .NET Foundation licenses this file to you under the MIT license.
     <MakeDir Directories="$([System.IO.Path]::GetDirectoryName($(NativeBinary)))" />
 
+    <PropertyGroup>
+      <EmccStackSize Condition="'$(EmccStackSize)' == ''">5MB</EmccStackSize>
```
This should be used to initialize `IlcWasmStackSize` below (if set). Note that our default is lower at 1MB (this is partly due to the fact we actually have two "stacks" in linear memory, so the usage is split between them, but also because 5MB seems too big for WASM, where in Release builds not that many things will go into the linear memory stack anyway).

maraf replied:
What is the difference between `-s STACK_SIZE` and `-s TOTAL_STACK`?

pavelsavara replied:
```
3.1.25 - 11/08/22
-----------------
- The `TOTAL_STACK` setting was renamed to `STACK_SIZE`.  The old name will
  continue to work as an alias. (#18128)
```

pavelsavara replied:
I think Mono needs large stack because when we interp tier up some of the nested generics of LINQ, the stack goes really deep. Hope I remember the right reason. Does NativeAOT run the unit test suite yet ?

SingleAccretion replied:
> Does NativeAOT run the unit test suite yet ?

Not really, it is being worked on though (#2455).

1MB is the stack size on Windows, where regular NAOT tests run (and succeed without SO). WASM stack usage in our codegen strategy will be as follows:
1) Native (LLVM) stack usage: _less_ than native because:
   - Some data that would usually be spilled is put into WASM locals instead.
   - Some data would go on the "shadow stack". These are things with GC pointers in them, as well as EH-live variables.
      - This "shadow stack" is basically the same as the native LLVM stack, but controlled by us. It's size is currently not configurable at 1MB as well.

So, from that I am pretty sure 1MB is more than necessary at this point.

maraf replied:
Ook, droped 5MB and pass `EmccStackSize` to `IlcWasmStackSize` if set by user

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -491,14 +492,17 @@ The .NET Foundation licenses this file to you under the MIT license.
+    <PropertyGroup>
+      <EmccStackSize Condition="'$(EmccStackSize)' == ''">5MB</EmccStackSize>
+      <EmccInitialHeapSize Condition="'$(EmccInitialHeapSize)' == '' and '$(_WasmCalculatedInitialHeapSize)' != '' and $(_WasmCalculatedInitialHeapSize) > 16777216">$(_WasmCalculatedInitialHeapSize)</EmccInitialHeapSize>
+      <EmccInitialHeapSize Condition="'$(EmccInitialHeapSize)' == ''">16777216</EmccInitialHeapSize>
```
```suggestion
```
Rationale: paired with the change below this will omit `-s INITIAL_MEMORY` from the default command line, which is the same as setting it to 16MB (current Emscripten default). More importantly, it means that once https://github.com/emscripten-core/emscripten/issues/20888 is fixed, we don't need to fix our targets.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -533,6 +537,13 @@ The .NET Foundation licenses this file to you under the MIT license.
       <CustomLinkerArg Condition="'$(IlcLlvmExceptionHandlingModel)' == 'cpp'" Include="-s DISABLE_EXCEPTION_CATCHING=0" />
+      
+      <CustomLinkerArg Include="$(EmccFlags)" />
+      <CustomLinkerArg Include="$(EmccLinkOptimizationFlag)" />
```
```suggestion
```
Rationale: I would like to omit this setting from the ones we support. It will break stack traces once we implement them, and we already offer optimization options that are better tested and well-supported: `'$(Optimize)' == 'true'` + `'$(NativeDebugSymbols)' == 'false'`.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -533,6 +537,13 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CustomLinkerArg Include="$(EmccFlags)" />
+      <CustomLinkerArg Include="$(EmccLinkOptimizationFlag)" />
+      <CustomLinkerArg Include="-s MAXIMUM_MEMORY=$(EmccMaximumHeapSize)" Condition="'$(EmccMaximumHeapSize)' != ''" />
+      <CustomLinkerArg Include="-s INITIAL_MEMORY=$(EmccInitialHeapSize)" />
```
```suggestion
      <CustomLinkerArg Include="-s INITIAL_MEMORY=$(EmccInitialHeapSize)" Condition="'$(EmccInitialHeapSize)' != ''" />
```

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -533,6 +537,13 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CustomLinkerArg Include="$(EmccLinkOptimizationFlag)" />
+      <CustomLinkerArg Include="-s MAXIMUM_MEMORY=$(EmccMaximumHeapSize)" Condition="'$(EmccMaximumHeapSize)' != ''" />
+      <CustomLinkerArg Include="-s INITIAL_MEMORY=$(EmccInitialHeapSize)" />
+      <CustomLinkerArg Include="-s STACK_SIZE=$(EmccStackSize)" />
```
```suggestion
```
Per the comment about `IlcWasmStackSize` above.

SingleAccretion commented on eng/liveBuilds.targets:
```diff
@@ -214,6 +216,22 @@
         $(LibrariesNativeArtifactsPath)*.dat;"
         IsNative="true" />
+      
+      <LibrariesRuntimeFiles Condition="'$(TargetOS)' == 'browser' and '$(RuntimeFlavor)' == 'CoreCLR'"
+                             Include="
+        $(CoreCLRNativeArtifactsPath)dotnet.js;
+        $(CoreCLRNativeArtifactsPath)dotnet.js.map;
+        $(CoreCLRNativeArtifactsPath)dotnet.runtime.js;
+        $(CoreCLRNativeArtifactsPath)dotnet.runtime.js.map;
+        $(CoreCLRNativeArtifactsPath)dotnet.d.ts;
+        $(CoreCLRNativeArtifactsPath)dotnet-legacy.d.ts;
+        $(CoreCLRNativeArtifactsPath)package.json;"
+        IsNative="true" />
+      <LibrariesRuntimeFiles Condition="'$(TargetOS)' == 'browser' and '$(RuntimeFlavor)' == 'CoreCLR'"
+                             Include="$(CoreCLRNativeArtifactsPath)src\es6\*.js"
+                             NativeSubDirectory="src\es6"
+                             IsNative="true" />
```
This treats the runtime JS as "libraries files", which works, and is fine for now, but is not ideal, since these are, after all, are "runtime" files. Could you add an item of "move runtime JS files to `/aotsdk` instead of `/framework`" to #2434?

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -491,14 +492,17 @@ The .NET Foundation licenses this file to you under the MIT license.
-      DependsOnTargets="$(LinkNativeDependsOn)">
+  <Import Project="$(MSBuildThisFileDirectory)DotNetJsApi.targets" Condition="'$(_targetArchitecture)' == 'wasm' and '$(DotNetJsApi)' == 'true'" />
 
+  <Target Name="SetupLlvmLinker">
```
Why was it required to split `LinkNativeLlvm`?

maraf replied:
Flags from `DotNetJsApi.targets` had to be applied after default ones. I don't remember exact details

maraf replied:
Here is a comparison if DotnetJs args are before/after default ones [dotnet.native.zip](https://github.com/dotnet/runtimelab/files/13671284/dotnet.native.zip). It's probably something with the JS library?

maraf replied:
Ook, removing `pal_random.lib.js` from linking when `DotnetJsApi` solved it (we have it in `crypto.ts`)

SingleAccretion replied:
It is good solution for the time being, but it shouldn't matter in principle, should it, since multiple `--js-library`s aren't supposed to override each other. It would be another item to add #2434: "investigate what happens when custom `--js-library`s are added by the user via `LinkerArg`".

maraf committed:
PrimaryRuntimeFlavor -> RuntimeFlavor

maraf committed:
Revert unintended TS changes

maraf committed:
Remove unused _WasmIntermediateOutputPath and _EmccLinkRsp

maraf committed:
EmccInitialHeapSize only if not empty

maraf committed:
EmccStackSize to IlcWasmStackSize

maraf committed:
Dont link pal_random.lib.js when dotnetjsapi. Revert LLVM linker target split

maraf committed:
Revert whitespace

SingleAccretion reviewed (approved):
The changes LGTM. Thank you!

yowl commented:
Very interesting, great work. Is it too early to add anything to the docs, like adding this to your project file if you want to use it

```
<DotNetJsApi>true</DotNetJsApi>
```

Are people going to think that `[JSImport]/[JSExport]` are supported with this change in NativeAOT-LLVM ?

maraf commented:
> Are people going to think that `[JSImport]/[JSExport]` are supported with this change in NativeAOT-LLVM ?

JS interop is tracked here https://github.com/dotnet/runtimelab/issues/2434. I'll make it work as separated PRs. All that's working now is (synchronously) calling Main

> Very interesting, great work. Is it too early to add anything to the docs, like adding this to your project file if you want to use it

I would wait a bit until there are more features working
